### ISSUE TITLE
Apply TrackCalib corrections to RDX run 2 MC sample

### ISSUE THREAD

**Issue by yipengsun** [umd-lhcb/lhcb-ntuples-gen#80]
We need to apply TrackCalib corrections to our MC samples.

## Concepts

- Tracking efficiency is defined as ![formula](https://render.githubusercontent.com/render/math?math=\epsilon%20=%20\frac{n_{\text{reconstructed}}}{n_{\text{reconstructable}}})
- The efficiency is simply _the probability that the trajectory of a charged particle that has passed-thru the full tracking system is reconstructed_
- The tracking efficiency **DOES NOT** account for differences between data and MC due to interactions with the material and decay-in-flight
- This can be estimated w/ MC
- In real data, the ![formula](https://render.githubusercontent.com/render/math?math=J/\psi%20\rightarrow%20\mu\mu) is used to obtain tracking efficiency
    - We use so-called "Tag-and-Prob method"
    - Basically, we fully reconstruct one Muon (call it "Tag"), then only partially reconstruct "Probe" so that the invariant mass of their common mother is known
    - Then we try to match the "Probe" to a track. If successful, it is considered "efficient". '
 - 3 efficiency measurement methods:
    
![3_track_eff_meas_methods](https://user-images.githubusercontent.com/33738176/126239269-47f24030-aac1-4a89-ba47-d820fcebd629.png)

## Useful links
- [TrackCalib TWiki](https://twiki.cern.ch/twiki/bin/view/LHCb/TrackCalib)
- [Tracking efficiency general, incl. Tag-and-Prob method and 3 methods](https://cds.cern.ch/record/1402577/files/LHCb-PUB-2011-025.pdf)

**Comment by yipengsun** [comment]
Trying to run `Urania` on `lxplus` failed:

```
> lb-run -c best Urania bash
WARNING:LbEnv.ProjectEnv.lookup:could not find Urania/prod in SearchPath([EnvSearchPathEntry('CMAKE_PREFIX_PATH', '/cvmfs/lhcb.cern.ch/lib/lhcb:/cvmfs/lhcb.cern.ch/lib/lcg/releases:/cvmfs/lhcb.cern.ch/lib/lcg/app/releases:/cvmfs/lhcb.cern.ch/lib/lcg/external:/cvmfs/lhcb.cern.ch/lib/contrib:/cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2129/stable/linux-64/lib/python3.8/site-packages/LbDevTools/data/cmake'), EnvSearchPathEntry('CMTPROJECTPATH', ''), EnvSearchPathEntry('LHCBPROJECTPATH', '')])
ERROR:lb-run:current host does not support any of Urania/prod available platforms
```

I first tried to run it according to its README:
```
> lb-run -c best --nightly lhcb-head Urania/HEAD bash
Traceback (most recent call last):
  File "/cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2129/stable/linux-64/lib/python3.8/site-packages/LbEnv/ProjectEnv/script.py", line 1130, in main
    super(LbRun, self).main()
  File "/cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2129/stable/linux-64/lib/python3.8/site-packages/xenv/__init__.py", line 289, in main
    self._makeEnv()
  File "/cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2129/stable/linux-64/lib/python3.8/site-packages/LbEnv/ProjectEnv/script.py", line 856, in _makeEnv
    self.opts.platform, self.opts.container = self._findPlatformAndContainer()
  File "/cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2129/stable/linux-64/lib/python3.8/site-packages/LbEnv/ProjectEnv/script.py", line 668, in _findPlatformAndContainer
    platforms = listPlatforms(self.project, self.version)
  File "/cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2129/stable/linux-64/lib/python3.8/site-packages/LbEnv/ProjectEnv/lookup.py", line 419, in listPlatforms
    return sorted(platforms, key=platform_sort_key, reverse=True)
  File "/cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2129/stable/linux-64/lib/python3.8/site-packages/LbEnv/ProjectEnv/lookup.py", line 69, in platform_sort_key
    comp = compilerKey(comp)
  File "/cvmfs/lhcb.cern.ch/lib/var/lib/LbEnv/2129/stable/linux-64/lib/python3.8/site-packages/LbEnv/ProjectEnv/lookup.py", line 92, in compilerKey
    family, version, extension = re.match(
AttributeError: 'NoneType' object has no attribute 'groups'
```

**Comment by yipengsun** [comment]
I tried the following instruction:
```shell
lb-dev --nightly lhcb-head Urania/v9r0
cd UrandiaDev_v9r0
git lb-use Urania
git lb-checkout Urania/master TrackCalib
git checkout Urania/master cmake  # this errors out
```

and got the following error:
```
error: pathspec 'cmake' did not match any file(s) known to git
```

**Comment by yipengsun** [comment]
Actually, without `-c`, the older version of `Urania` seems to work:
```
> lb-run Urania/v9r0 bash
WARNING:lb-run:Decided best container to use is None
[Urania v9r0] UraniaDev_v9r0 $ python TrackCalib/scripts/CreateTrackEffTables.py "2016_25ns"
```

This seems to run fine.

**Comment by yipengsun** [comment]
No. It's a false hope. The fitting script stuck on some pop-up plots, which according to the TWiki was fixed. I then tried to manually clone the latest `master` branch of Urania and run the TrackCalib portion within Uranian/v9r0, but I got the following error:
```
❯ lb-run Urania/v9r0 python TrackCalib/scripts/CreateTrackEffTables.py "2015_25ns" 
WARNING:lb-run:Decided best container to use is None

RooFit v3.60 -- Developed by Wouter Verkerke and David Kirkby 
                Copyright (C) 2000-2013 NIKHEF, University of California & Stanford University
                All rights reserved, please read http://roofit.sourceforge.net/license.txt


##########################################################
#               Welcome to TrackEffReducer               #
#                    Running on MC                       #
##########################################################
connect localhost port 6011: Connection refused
```

Don't know what will happen or if the session is already stuck. 

Anyway, I still don't see a way of running TrackCalib. The instruction really needs some updates. Luckily Veronica has got back to be (a Run1-2 person) and hopefully I can get some feedback soon.

**Comment by yipengsun** [comment]
The Urania README is at: https://gitlab.cern.ch/lhcb/Urania
The TrackCalib TWiki is at: https://twiki.cern.ch/twiki/bin/view/LHCb/TrackCalib

**Comment by yipengsun** [comment]
The following TWiki seems to provide the static efficiency tables:
https://twiki.cern.ch/twiki/bin/viewauth/LHCbInternal/LHCbTrackingEfficiencies

**Comment by yipengsun** [comment]
The 2016 tracking weights (it is not differentiated by polarity!):

| η \ p   | 0 (U)       | 1 (5000.0)   | 2 (10000.0)   | 3 (20000.0)   | 4 (40000.0)   | 5 (100000.0)   | 6 (O)       |
|---------|-------------|--------------|---------------|---------------|---------------|----------------|-------------|
| 0 (U)   | 0.00 ± 0.00 | 0.00 ± 0.00  | 0.00 ± 0.00   | 0.00 ± 0.00   | 0.00 ± 0.00   | 0.00 ± 0.00    | 0.00 ± 0.00 |
| 1 (1.9) | 0.00 ± 0.00 | 0.96 ± 0.00  | 0.99 ± 0.00   | 1.00 ± 0.00   | 0.99 ± 0.00   | 1.00 ± 0.05    | 0.00 ± 0.00 |
| 2 (3.2) | 0.00 ± 0.00 | 1.00 ± 0.05  | 1.00 ± 0.00   | 0.99 ± 0.00   | 1.00 ± 0.05   | 1.00 ± 0.00    | 0.00 ± 0.00 |
| 3 (O)   | 0.00 ± 0.00 | 0.00 ± 0.00  | 0.00 ± 0.00   | 0.00 ± 0.00   | 0.00 ± 0.00   | 0.00 ± 0.00    | 0.00 ± 0.00 |

**Comment by yipengsun** [comment]
Tracking weights for `D*`:

- `Mu`: 0.8855
- `K`: 0.9868
- `Pi`: 0.9854
- `slow Pi`: 0.4434

For `D0`:

- `Mu`: 0.8734
- `K`: 0.9822
- `Pi`: 0.9827

Previously I have validated that the weight applying tool is working properly. Here the main concern is if the `P` and `ETA` are swapped. Obviously they are not, because in that case we'll get lots of 0s.

A slight concern is the `slow Pi`.

**Comment by yipengsun** [comment]
I updated the histo printer and now the full bin boundaries are printed out. For the first bin, the lower bound and upper bound are both printed; for the rest of the bin, only the upper bound is printed.

| η \ p       | 0 (U)       | 1 (5000.0,10000.0)   | 2 (20000.0)   | 3 (40000.0)   | 4 (100000.0)   | 5 (200000.0)   | 6 (O)       |
|-------------|-------------|----------------------|---------------|---------------|----------------|----------------|-------------|
| 0 (U)       | 0.00 ± 0.00 | 0.00 ± 0.00          | 0.00 ± 0.00   | 0.00 ± 0.00   | 0.00 ± 0.00    | 0.00 ± 0.00    | 0.00 ± 0.00 |
| 1 (1.9,3.2) | 0.00 ± 0.00 | 0.96 ± 0.00          | 0.99 ± 0.00   | 1.00 ± 0.00   | 0.99 ± 0.00    | 1.00 ± 0.05    | 0.00 ± 0.00 |
| 2 (4.9)     | 0.00 ± 0.00 | 1.00 ± 0.05          | 1.00 ± 0.00   | 0.99 ± 0.00   | 1.00 ± 0.05    | 1.00 ± 0.00    | 0.00 ± 0.00 |
| 3 (O)       | 0.00 ± 0.00 | 0.00 ± 0.00          | 0.00 ± 0.00   | 0.00 ± 0.00   | 0.00 ± 0.00    | 0.00 ± 0.00    | 0.00 ± 0.00 |

**Comment by yipengsun** [comment]
OK, I plotted the P-ETA 2D heat maps for all of our final state particles (`K, Pi, Mu, slow Pi`):

## `D0` tree
![D0_k_p_eta](https://user-images.githubusercontent.com/33738176/146422091-d55173c2-91df-4544-836d-1763481e85a2.png)
![D0_mu_p_eta](https://user-images.githubusercontent.com/33738176/146422094-6e8788a9-21e2-40f5-bc69-161f5913fc0e.png)
![D0_pi_p_eta](https://user-images.githubusercontent.com/33738176/146422098-a36912d2-a997-409b-8604-faab1fca93b5.png)

## `D*` tree
![Dst_k_p_eta](https://user-images.githubusercontent.com/33738176/146422102-cf08707e-50a9-4757-a78c-7c356283fd0a.png)
![Dst_mu_p_eta](https://user-images.githubusercontent.com/33738176/146422105-e78d6f44-c362-48a2-8770-13968883bba4.png)
![Dst_pi_p_eta](https://user-images.githubusercontent.com/33738176/146422107-2c677e7e-27a8-43bf-9771-ca1619452873.png)
![Dst_spi_p_eta](https://user-images.githubusercontent.com/33738176/146422110-e4fba8d1-6140-43dc-92d8-4d11a26eeeff.png)

**Comment by yipengsun** [comment]
I've followed Phoebe's suggestion to make sure the under/overflow bins are adjusted to the nearest bins.

I've also updated the plots above, and as you can see, the efficiencies are now close to unity.

**Comment by yipengsun** [comment]
Need to double check the `slow Pi` correction as 50% of samples are outside the official range.

**Comment by yipengsun** [comment]
@manuelfs I've noted in our doc that we should be using `+/- 0.05` for uncertainty for the bins that are not covered by the official tracking sample: https://github.com/umd-lhcb/rdx-run2-analysis/blob/master/docs/reweight/pid_tracking_reweight.md#tracking

If you think everything's documented, feel free to close this issue.
